---
title: "ST QC"
author: "Eva, Linnea and Zaneta"
date: "6/20/2019"
output: xaringan::moon_reader
---



# Spatial Transcriptomics Data


- RNA-seq based method 
- No individual cells -> spots with probes under tissue 
  - 1 spot ~ 30 cells
  - + Spatial info (coordinates)
  
---

# ST DATA


Sequencing -> QFilter, Mapping and Annotation -> .tsv files

 - rownames:Ensembl gene names
 - colnames:Spot coordinates

 
Spots_under_tissueselection matrix -> .tsv file spot coordinates of spots under tissue

---
# First part of ST.QC


```{r, echo=FALSE}
library(magrittr)
library(dplyr)
library(ggplot2)
library(devtools)
library(usethis)
library(roxygen2)
library(knitr)
library(rmarkdown)
library(reshape2)
library(Rcpp)
library(data.table)
library(Matrix)
library(gridExtra)

tsv_file <- "/Users/evagracia/AZKManuscript/tsv files/190415_Kidney_EXP1_NHK1_D2_stdata.tsv"
alignment_file <- "/Users/evagracia/AZKManuscript/selection files/spot_data-selection-EXP1_NHK1_D2.tsv"


S1 <- t(data.frame(fread(input = tsv_file, sep = "\t", header = T), row.names = 1))

S1_alignment <- read.table(file = alignment_file, header = T, stringsAsFactors = F)
S1[1:4,1:4]


rownames(S1_alignment) <- paste(S1_alignment$x, S1_alignment$y, sep = "x")
S1_alignment[1:4,1:6]
```

---

# Filtering and Subsetting the Data


```{r, echo=FALSE}


rownames(S1_alignment) <- paste(S1_alignment$x, S1_alignment$y, sep = "x")

subset_data <- function(data, alignment) {

  data_subset <- data[, intersect(colnames(data), rownames(alignment))]
  

  data_subset <- data_subset[rowSums(data_subset) > 0, ]
  

  remove_genes <- grep(pattern = "ambiguous", x = rownames(data_subset), value = T)
  data_subset <- data_subset[!rownames(data_subset) %in% remove_genes, ]
  

  cat("Number of genes in raw dataset: ", nrow(data), "\n")
  cat("Number of genes in subset dataset: ", nrow(data_subset), "\n")
  cat("Number of spots in raw dataset: ", ncol(data), "\n")
  cat("Number of spots in subset dataset: ", ncol(data_subset), "\n\n")
  return(data_subset)
}


S1_subset <- subset_data(S1, S1_alignment)

```
---

# Converting gene ids and sorting


```{r, echo=FALSE}


ensids <- read.table(file = "/Users/evagracia/AZKManuscript/genes.tsv", header = T, stringsAsFactors = T)
rownames(ensids) <- ensids$gene_id


accepted_genes <- ensids$gene_id[ensids$gene_type %in% c("protein_coding", "lincRNA", "IG_C_gene")]

rownames(S1_subset) <- ensids[rownames(S1_subset), ]$gene_name


remove_genes_S1 <- c(grep(pattern = "^MTRNR", value = T, x = rownames(S1_subset)), "MALAT1")
S1_subset <- S1_subset[!rownames(S1_subset) %in% remove_genes_S1, ]
S1_subset[1:15,1:5]
```
---

# Histo_Unique Function


```{r, echo=FALSE}
S1_df <- as.data.frame(S1_subset)
unique.genes <- apply(S1_df, 2, function(x) sum(x > 0))
unique.transcripts <- colSums(S1_df)

gg <- data.frame(unique.genes, unique.transcripts)
gg <- add_rownames(gg, var = "Spot")

a <- ggplot(gg, aes (x = unique.genes))+
  geom_histogram(bins = 100, color = "blue")

b <- ggplot(gg, aes (x = unique.transcripts))+
  geom_histogram(bins = 100, color = "red")
grid.arrange(a,b)

```
---

# Final Thoughts


![Final](horse.jpg)

